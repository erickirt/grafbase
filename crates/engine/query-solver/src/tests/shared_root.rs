use crate::assert_solving_snapshots;

const SCHEMA: &str = r###"
enum join__Graph {
  CATEGORY @join__graph(name: "category", url: "http://localhost:4200/shared-root/category")
  NAME @join__graph(name: "name", url: "http://localhost:4200/shared-root/name")
  PRICE @join__graph(name: "price", url: "http://localhost:4200/shared-root/price")
}

type Product
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
  nested: Nested!
  category: String @join__field(graph: CATEGORY)
  name: String @join__field(graph: NAME)
  price: Float @join__field(graph: PRICE)
}

type Nested
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
}

type Query
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  product: Product
  products: [Product]
}
"###;

#[tokio::test]
async fn all_fields() {
    assert_solving_snapshots!(
        "all_fields",
        SCHEMA,
        r#"
        query {
          products {
            price
            category
            id
            name
          }
        }
        "#
    );
}

#[tokio::test]
async fn nested_id_field_with_all_fields() {
    assert_solving_snapshots!(
        "nested_id_with_all_fields",
        SCHEMA,
        r#"
        query {
          products {
            price
            category
            name
            nested { id }
          }
        }
        "#
    );
}

#[tokio::test]
async fn single_field() {
    assert_solving_snapshots!(
        "single_field",
        SCHEMA,
        r#"
        query {
          products {
            price
          }
        }
        "#
    );
}
