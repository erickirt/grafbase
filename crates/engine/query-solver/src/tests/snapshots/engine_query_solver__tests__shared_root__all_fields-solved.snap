---
source: crates/engine/query-solver/src/tests/shared_root.rs
expression: "digraph {\n    0 [ label = \"root\", color=forestgreen ]\n    1 [ label = \"Product.price\", color=forestgreen ]\n    2 [ label = \"Product.category\", color=forestgreen ]\n    3 [ label = \"Product.id\", color=forestgreen ]\n    4 [ label = \"Product.name\", color=forestgreen ]\n    5 [ label = \"Root#category\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    6 [ label = \"products#category\", shape=box, color=dodgerblue, color=forestgreen ]\n    7 [ label = \"Root#name\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    8 [ label = \"products#name\", shape=box, color=dodgerblue, color=forestgreen ]\n    9 [ label = \"Root#price\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    10 [ label = \"products#price\", shape=box, color=dodgerblue, color=forestgreen ]\n    11 [ label = \"price#price\", shape=box, color=dodgerblue, color=forestgreen ]\n    12 [ label = \"id#price\", shape=box, color=dodgerblue, color=forestgreen ]\n    13 [ label = \"id#name\", shape=box, color=dodgerblue, style=dashed ]\n    14 [ label = \"name#name\", shape=box, color=dodgerblue, color=forestgreen ]\n    15 [ label = \"category#category\", shape=box, color=dodgerblue, color=forestgreen ]\n    16 [ label = \"id#category\", shape=box, color=dodgerblue, style=dashed ]\n    17 [ label=\"\", style=dashed]\n    0 -> 5 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    5 -> 6 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    0 -> 7 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    7 -> 8 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    0 -> 9 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    9 -> 10 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    10 -> 11 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    11 -> 1 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    10 -> 12 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    12 -> 3 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    8 -> 13 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    13 -> 3 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    8 -> 14 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    14 -> 4 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    6 -> 15 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    15 -> 2 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    6 -> 16 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    16 -> 3 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    17 -> 0 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n}\n"
---
digraph {
    0 [ label = "root", steiner=1 ]
    1 [ label = "Product.price", steiner=1 ]
    2 [ label = "Product.category", steiner=1 ]
    3 [ label = "Product.id", steiner=1 ]
    4 [ label = "Product.name", steiner=1 ]
    5 [ label = "Root#category", steiner=1 ]
    6 [ label = "products#category", steiner=1 ]
    7 [ label = "Root#name", steiner=1 ]
    8 [ label = "products#name", steiner=1 ]
    9 [ label = "Root#price", steiner=1 ]
    10 [ label = "products#price", steiner=1 ]
    11 [ label = "price#price", steiner=1 ]
    12 [ label = "id#price", steiner=1 ]
    13 [ label = "id#name", steiner=0 ]
    14 [ label = "name#name", steiner=1 ]
    15 [ label = "category#category", steiner=1 ]
    16 [ label = "id#category", steiner=0 ]
    17 [ label="", style=dashed]
    0 -> 5 [ cost=0, steiner=1]
    5 -> 6 [ cost=0, steiner=1]
    0 -> 7 [ cost=0, steiner=1]
    7 -> 8 [ cost=0, steiner=1]
    0 -> 9 [ cost=0, steiner=1]
    9 -> 10 [ cost=0, steiner=1]
    10 -> 11 [ cost=0, steiner=1]
    11 -> 1 [ cost=0, steiner=1]
    10 -> 12 [ cost=0, steiner=1]
    12 -> 3 [ cost=0, steiner=1]
    8 -> 13 [ cost=0, steiner=0]
    13 -> 3 [ cost=0, steiner=0]
    8 -> 14 [ cost=0, steiner=1]
    14 -> 4 [ cost=0, steiner=1]
    6 -> 15 [ cost=0, steiner=1]
    15 -> 2 [ cost=0, steiner=1]
    6 -> 16 [ cost=0, steiner=0]
    16 -> 3 [ cost=0, steiner=0]
    17 -> 0 [ cost=0, steiner=0]
}
